### PR TITLE
[doc] Update SIL.rst to reflect debug_value_addr deprecation

### PR DESCRIPTION
**Pull request by mshockwave**
And restructure the "Debug Information" section to make it flow better.

### CODE REVIEW

File: docs/SIL.rst [hunk 1]
```diff
@@ -3556,23 +3581,13 @@ like AST nodes or strings. Here is an example::
...
-
-This indicates that the value of a declaration with address-only type
-has changed value to the specified operand.  The declaration in
-question is identified by the SILLocation attached to the
-debug_value_addr instruction.
+In the snippet above, source variable "the_struct" has an aggregate type ``$MyStruct`` and we use di-expression with ``op_fragment`` operator to associate ``%1`` to the ``y`` member variable (via the ``#MyStruct.y`` directive) inside "the_struct".
+Note that the extra source location directive follows rigt after ``name "the_struct"`` indicate that "the_struct" was originally declared in line 8, but not until line 9 -- the current ``debug_value`` instruction's source location -- does member ``y`` got updated with SSA value ``%1``.
 
-Note that this instruction can be replaced by ``debug_value`` + di-expression operator that is equivalent to LLVM's ``DW_OP_deref``.
+It is worth noting that Di-expression in SIL is similar to ``!DIExpression`` in LLVM.
```
Comment (adrian-prantl) [inline docs/SIL.rst:3587]: A SIL DI-expression is similar to a ``DIExpression`` in LLVM IR debug info metadata (https://www.llvm.org/docs/LangRef.html#diexpression). While LLVM DIExpressions are a list of 64-bit integers, SIL DI-expressions can have ...

**Comment by adrian-prantl** [comment]
Thanks! Looking good, I just have a few minor wording suggestions inside

File: docs/SIL.rst [hunk 2]
```diff
@@ -3507,47 +3506,73 @@ The operand must have loadable type.
...
-value with the source variable in an indirect fashion. For example,
-we can use the ``op_fragment`` operator to specify that the SSA value
-is originated from a struct field inside the source variable (which has
-an aggregate data type). Di-expression in SIL works similarly to LLVM's
-``!DIExpression`` metadata. Where both of them adopt a stack based
-execution model to evaluate the expression. The biggest difference between
-them is that LLVM always represent ``!DIExpression`` elements as 64-bit
-integers, while SIL's di-expression can have elements with various types,
-like AST nodes or strings. Here is an example::
+value with the source variable in an indirect fashion. Di-expression in SIL
```
[the remark this answers is not in the document]
Comment (mshockwave) [reply docs/SIL.rst:3544]: Yeah that's a good idea. Though I used "di-expression" in many code comments but let's change them here first.

**Comment by mshockwave** [comment]
@swift-ci please test

**Review by adrian-prantl — APPROVED** [review]

**Pull request merged**